### DISCUSSION TITLE
Is this the correct way to implement Realtime-kt?

### DISCUSSION THREAD

**Discussion by premjit-develops** [supabase-community/supabase-kt#597] [Q&A]
I'm trying to integrate real-time functionality into my app using the chat demo as a reference, but I'm struggling to do so correctly. Can I receive guidance on implementing real-time functionality into my app?

**DI** 

    @Provides
    @Singleton
    fun providesRealtime(client: SupabaseClient): RealtimeChannel = client.realtime.channel("orders")

**Cart Screen**
```
 LaunchedEffect(Unit) {
    
        viewModel.connectToRealtime(this)
      
    }
```

**Cart ViewModel** 

    fun connectToRealtime(scope: CoroutineScope) = viewModelScope.launch{

        try {

           // val realtimeChannel = realtime.channel("orders")

            val changeFlow = realtimeChannel.postgresChangeFlow<PostgresAction>(schema = "public") {
                table = "orders"
            }
            changeFlow.onEach {
                when (it) {
                    is PostgresAction.Delete -> {
                        val updatedOrder = it.decodeOldRecord<Order>()
                        val updatedOrderList = _orderLines.value.orderList + updatedOrder
                        _orderLines.value = _orderLines.value.copy(
                            orderList = updatedOrderList
                        )

                        Log.d(TAG, "DeleteItem : ${it.oldRecord}")

                    }

                    is PostgresAction.Insert -> {
                        val updatedOrder = it.decodeRecord<Order>()
                        val updatedOrderList = _orderLines.value.orderList + updatedOrder
                        _orderLines.value = _orderLines.value.copy(
                            orderList = updatedOrderList
                        )

                        Log.d(TAG, "InsertItem : ${it.record}")
                    }

                    is PostgresAction.Select -> {
                        val updatedOrder = it.decodeRecord<Order>()
                        val updatedOrderList = _orderLines.value.orderList + updatedOrder
                        _orderLines.value = _orderLines.value.copy(
                            orderList = updatedOrderList
                        )

                        Log.d(TAG, "SelectedItem : ${it.record}")
                    }

                    is PostgresAction.Update -> {
                        val updatedOrder = it.decodeRecord<Order>()
                        val updatedOrderList = _orderLines.value.orderList + updatedOrder
                        _orderLines.value = _orderLines.value.copy(
                            orderList = updatedOrderList
                        )

                        Log.d(TAG, "UpdateItem : ${it.record}")
                    }

                }
            }.launchIn(scope)

            realtimeChannel.subscribe()
        }catch (e: RestException){
            e.printStackTrace()
        }
    }
    
    
 **Cart UI State**

```
data class CartUiState(
    val orderList: List<Order> = listOf()
)
```

**Comment by jan-tennert** [comment]
> fun connectToRealtime(scope: CoroutineScope) = viewModelScope.launch{

I wouldn't use the scope from `LaunchedEffect`, you are already in the `viewModelScope` so you can use it. You can also take a look at a new API which doesn't require channels: https://supabase.com/docs/reference/kotlin/stream

**Comment by premjit-develops** [comment]
I forgot to add getting this error:
  ```
java.lang.IllegalStateException: You cannot call postgresChangeFlow after joining the channel
                 	at com.foodz.ui.feature.home.cart.CartVM$connectToRealtime$1.invokeSuspend(CartVM.kt:239)
                 	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
                 	at kotlinx.coroutines.internal.DispatchedContinuationKt.resumeCancellableWith(DispatchedContinuation.kt:363)
                 	at kotlinx.coroutines.intrinsics.CancellableKt.startCoroutineCancellable(Cancellable.kt:26)
                 	at kotlinx.coroutines.intrinsics.CancellableKt.startCoroutineCancellable$default(Cancellable.kt:21)
                 	at kotlinx.coroutines.CoroutineStart.invoke(CoroutineStart.kt:88)
                 	at kotlinx.coroutines.AbstractCoroutine.start(AbstractCoroutine.kt:123)
                 	at kotlinx.coroutines.BuildersKt__Builders_commonKt.launch(Builders.common.kt:52)
                 	at kotlinx.coroutines.BuildersKt.launch(Unknown Source:1)
                 	at kotlinx.coroutines.BuildersKt__Builders_commonKt.launch$default(Builders.common.kt:43)
                 	at kotlinx.coroutines.BuildersKt.launch$default(Unknown Source:1)
                 	at com.foodz.ui.feature.home.cart.CartVM.connectToRealtime(CartVM.kt:167)

**Reply by jan-tennert** [reply]
You might be calling subscribe before calling this method or reusing a subscribed channel. Can't see much in your snippet.

**Comment by premjit-develops** [comment]
Maybe this ViewModel implementation part can help to identify this error

```

@HiltViewModel
class CartVM @Inject constructor(
    val authRepository: AuthRepository,
    val repository: DatabaseRepository,
    val realtimeChannel: RealtimeChannel
) : ViewModel() {


    private val _orderLines = MutableStateFlow(CartUiState())
    val uiState: StateFlow<CartUiState> = _orderLines.asStateFlow()

    val userId = authRepository.isCurrentUser()?.id.orEmpty()

    init {
        getAllOrders()

    }

    fun getAllOrders() = viewModelScope.launch {
        repository.getAllOrders(userId).collect {
            when (it) {
                is ResultState.Loading -> {
                    _orderLines.value = _orderLines.value.copy(
                        isLoading = true,
                    )
                }

                is ResultState.Success -> {
                    _orderLines.value = _orderLines.value.copy(
                        orderList = it.data,
                        isLoading = false
                    )
                }

                is ResultState.Error -> {
                    _orderLines.value = _orderLines.value.copy(
                        isLoading = false,
                        isError = true
                    )
                }
            }
        }
    }
    
       fun connectToRealtime() = viewModelScope.launch {

        try {

            // val realtimeChannel = realtime.channel("orders")

            val changeFlow = realtimeChannel.postgresChangeFlow<PostgresAction>(schema = "public") {
                table = "orders"
            }
            changeFlow.onEach {
                when (it) {
                    is PostgresAction.Delete -> {
                        val updatedOrder = it.decodeOldRecord<Order>()
                        val updatedOrderList = _orderLines.value.orderList + updatedOrder
                        _orderLines.value = _orderLines.value.copy(
                            orderList = updatedOrderList
                        )

                        Log.d(TAG, "DeleteItem : ${it.oldRecord}")

                    }

                    is PostgresAction.Insert -> {
                        val updatedOrder = it.decodeRecord<Order>()
                        val updatedOrderList = _orderLines.value.orderList + updatedOrder
                        _orderLines.value = _orderLines.value.copy(
                            orderList = updatedOrderList
                        )

                        Log.d(TAG, "InsertItem : ${it.record}")
                    }

                    is PostgresAction.Select -> {
                        val updatedOrder = it.decodeRecord<Order>()
                        val updatedOrderList = _orderLines.value.orderList + updatedOrder
                        _orderLines.value = _orderLines.value.copy(
                            orderList = updatedOrderList
                        )

                        Log.d(TAG, "SelectedItem : ${it.record}")
                    }

                    is PostgresAction.Update -> {
                        val updatedOrder = it.decodeRecord<Order>()
                        val updatedOrderList = _orderLines.value.orderList + updatedOrder
                        _orderLines.value = _orderLines.value.copy(
                            orderList = updatedOrderList
                        )

                        Log.d(TAG, "UpdateItem : ${it.record}")
                    }

                }
            }

            realtimeChannel.subscribe()
        } catch (e: RestException) {
            e.printStackTrace()
        }
    }
}


```

**Reply by jan-tennert** [reply]
You are missing a launchIn on your flow:
```kotlin
fun connectToRealtime() = viewModelScope.launch {

        try {

            // val realtimeChannel = realtime.channel("orders")

            val changeFlow = realtimeChannel.postgresChangeFlow<PostgresAction>(schema = "public") {
                table = "orders"
            }
            changeFlow.onEach {
                when (it) {
                    is PostgresAction.Delete -> {
                        val updatedOrder = it.decodeOldRecord<Order>()
                        val updatedOrderList = _orderLines.value.orderList + updatedOrder
                        _orderLines.value = _orderLines.value.copy(
                            orderList = updatedOrderList
                        )

                        Log.d(TAG, "DeleteItem : ${it.oldRecord}")

                    }

                    is PostgresAction.Insert -> {
                        val updatedOrder = it.decodeRecord<Order>()
                        val updatedOrderList = _orderLines.value.orderList + updatedOrder
                        _orderLines.value = _orderLines.value.copy(
                            orderList = updatedOrderList
                        )

                        Log.d(TAG, "InsertItem : ${it.record}")
                    }

                    is PostgresAction.Select -> {
                        val updatedOrder = it.decodeRecord<Order>()
                        val updatedOrderList = _orderLines.value.orderList + updatedOrder
                        _orderLines.value = _orderLines.value.copy(
                            orderList = updatedOrderList
                        )

                        Log.d(TAG, "SelectedItem : ${it.record}")
                    }

                    is PostgresAction.Update -> {
                        val updatedOrder = it.decodeRecord<Order>()
                        val updatedOrderList = _orderLines.value.orderList + updatedOrder
                        _orderLines.value = _orderLines.value.copy(
                            orderList = updatedOrderList
                        )

                        Log.d(TAG, "UpdateItem : ${it.record}")
                    }

                }
            }.launchIn(this) //otherwise you will get no updates

            realtimeChannel.subscribe()
        } catch (e: RestException) {
            e.printStackTrace()
        }
    }
```
But that doesn't explain the error. Is this method really the only instance where you subscribe to the channel? Can you print to the console to see if this method is somehow called multiple times? Also add some logs before and after calling `postgresChangeFlow`

**Reply by premjit-develops** [reply]
I'm trying to get real-time updates for my cart when the user adds new food to the cart screen
```
 fun connectToRealtime() = viewModelScope.launch{

        try {
            Log.d(TAG, "Connecting to realtime...")

          
            Log.d(TAG, "Before calling postgresChangeFlow")

            val changeFlow = realtimeChannel.postgresChangeFlow<PostgresAction>(schema = "public") {
                table = "orders"
            }

            Log.d(TAG, "After calling postgresChangeFlow")

            changeFlow.onEach {
                when (it) {
                    is PostgresAction.Delete -> {
                        val updatedOrder = it.decodeOldRecord<Order>()
                        val updatedOrderList = _orderLines.value.orderList + updatedOrder
                        _orderLines.value = _orderLines.value.copy(
                            orderList = updatedOrderList,
                        )

                        Log.d(TAG, "DeleteItem : $updatedOrder")

                    }

                    is PostgresAction.Insert -> {
                        val updatedOrder = it.decodeRecord<Order>()
                        val updatedOrderList = _orderLines.value.orderList + updatedOrder
                        _orderLines.value = _orderLines.value.copy(
                            orderList = updatedOrderList,
                        )

                        Log.d(TAG, "InsertItem : $updatedOrder")
                    }

                    is PostgresAction.Select -> {
                        val updatedOrder = it.decodeRecord<Order>()
                        val updatedOrderList = _orderLines.value.orderList + updatedOrder
                        _orderLines.value = _orderLines.value.copy(
                            orderList = updatedOrderList,
                        )

                        Log.d(TAG, "SelectedItem : ${it.record}")
                    }

                    is PostgresAction.Update -> {
                        val updatedOrder = it.decodeRecord<Order>()
                        val updatedOrderList = _orderLines.value.orderList + updatedOrder
                        _orderLines.value = _orderLines.value.copy(
                            orderList = updatedOrderList,
                        )

                        Log.d(TAG, "UpdateItem : ${it.record}")
                    }

                }
            }.launchIn(this)

            realtimeChannel.subscribe()
        } catch (e: RestException) {
            e.printStackTrace()

        }catch (e: Exception){
            e.printStackTrace()
        }

    }
    
```

When navigating to the cart screen for the first time

![s1](https://github.com/supabase-community/supabase-kt/assets/72644801/5286a7d5-82fa-4f8e-a6f1-35c41ccb6da3)

 if I go there again and then delete or insert something, I get a serialization error.
![s2](https://github.com/supabase-community/supabase-kt/assets/72644801/63943066-eeae-47ac-9ffe-9b01579e21ab)

![s3](https://github.com/supabase-community/supabase-kt/assets/72644801/354a0d1c-daa6-43bb-a82b-726ec2c37a98)


My Order class

```
@Immutable
@Serializable
data class Order(
    val id: String,

    @SerialName("user_id")
    val userId: String,

    @SerialName("order_total")
    val orderTotal: Double,

    @SerialName("delivery_status")
    val deliveryStatus: String,

    @SerialName("food")
    val foodId: ShortFoodData,

    val quantity: Int
)

@Immutable
@Serializable
data class ShortFoodData(
    val id: Long,
    val name: String,
    val imageUrl: String,
    val price: Double,
    val tagline: String,
)
```

**Reply by jan-tennert** [reply]
So I assume you get the realtime error because you don't unsubscribe after leaving the screen and then try to call this `postgresChangeFlow` again after renavigating to this screen. For the serialization error, Supabase doesn't include all fields in the `DELETE` event, so `it.decodeOldRecord<Order>()` fails as there is only the primary key available.

**Reply by premjit-develops** [reply]
Yes, you're correct that IllegalStateException arises from not setting unsubscribe. However, I'm facing a serialization error. When I directly log the items, they print correctly. But when I add them to the order list, I encounter a serialization error. How can I resolve this?
```
val updatedOrder = it.decodeRecord<Order>()
                        val updatedOrderList = _orderLines.value.orderList + updatedOrder
                        _orderLines.value = _orderLines.value.copy(
                            orderList = updatedOrderList,
                        )
                        
                            Log.d(TAG, "InsertItem : ${it.record}")
```

**Reply by jan-tennert** [reply]
For which event? For the delete event? Or for all?

**Reply by premjit-develops** [reply]
Mainly delete and insert.

**Reply by jan-tennert** [reply]
Are you sure that happens on the delete event? Does the printed record contain all fields? And as I said 
> Supabase doesn't include all fields in the DELETE event, so it.decodeOldRecord<Order>() fails as there is only the primary key available.

**Reply by premjit-develops** [reply]
If I print "Insert" or "Delete" along with decodeRecord, it results in a serialization error.
```
 val updatedOrder = it.decodeRecord<Order>()

 Log.d(TAG, "InsertItem : $updatedOrder")
```

If I were to directly print the JsonObject from a PostgresAction insertion

```
is PostgresAction.Insert -> {
  
     Log.d(TAG, "InsertItem : ${it.record}")
}
```
Output
![Screenshot from 2024-05-17 21-19-55](https://github.com/supabase-community/supabase-kt/assets/72644801/9cbe9c3a-c19a-4017-b850-00437d02080c)

**Reply by premjit-develops** [reply]
I've tested `selectAsFlow` and `PostgresAction` without nested model classes, and they work as expected. However, when I use a nested class like `Order` with `ShortFoodData` embedded, a serialization error occurs, indicating that `food_id` is not found. But when I use the same `Order` class with the `select` method `Columns.raw("*, food:foods(*)")`, everything works as expected.


Error: `kotlinx.serialization.MissingFieldException: Field 'food' is required for type with serial name 'com.foodz.data.model.Order', but it was missing at path: $[0]`

**Reply by jan-tennert** [reply]
Yes, the `selectAsFlow` method does not have a columns parameter due to how realtime works, so then you probably have to use your previous approach

**Reply by premjit-develops** [reply]
I tried using the modified `postgresListDataFlow` extension function, including columns, and it works. However, when I implement it with

`realtimeChannel.subscribe() 
emitAll(data) `
  
it is not working only deletion works in real-time. If I try to insert something, I encounter the same `MissingFieldError`
```
inline fun <reified Data : Any, Value> RealtimeChannel.realtimeListDataFlow(
    schema: String = "public",
    table: String,
    filter: FilterOperation? = null,
    columns: Columns,
    primaryKey: KProperty1<Data, Value>,
): Flow<List<Data>> = realtimeListDataFlow<Data>(
    filter = filter,
    table = table,
    schema = schema,
    columns = columns,
    primaryKey = PrimaryKey(
        primaryKey.name
    ){
        primaryKey.get(it).toString()
    }
)
```

**Reply by jan-tennert** [reply]
Does your datatype expect some joined fields? Realtime doesn't provide referenced objects in their payloads. Only the raw table row.

**Reply by premjit-develops** [reply]
Yes, some joined fields expected

**Reply by jan-tennert** [reply]
You have to listen to each table's changes separately. That's what I'm doing for my apps, fetch & listen to updates separately, store them in a local sql db and join them via requests to the local db.

**Reply by premjit-develops** [reply]
Currently not possible with Realtime apis?

**Reply by jan-tennert** [reply]
Yes, realtime only supports listening to tables separately. You'll have to join them yourself.

**Reply by premjit-develops** [reply]
I'm saying that direct join like `select()` columns is it possible currently with realtime apis or in furture?

**Reply by jan-tennert** [reply]
No, that's not possible, and I don't know when or if they implement something like this, same with multiple filters.
See https://github.com/supabase/realtime/issues/303.

**Reply by premjit-develops** [reply]
Oh..that means Supabase Realtime Engine currently doesn't support join queries. By the way, thanks for the replies.
